hatchery/authz.go

```diff
@@ -0,0 +1,275 @@
+package hatchery
+
+import (
+	"bytes"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"log"
+	"net/http"
+	"time"
+)
+
+type AuthzConfig struct {
+	Version          float32 `json:"version"`
+	AuthzVersion_0_1 AuthzVersion_0_1
+}
+
+type AuthzVersion_0_1 struct {
+	And           []AuthzVersion_0_1 `json:"and"`
+	Or            []AuthzVersion_0_1 `json:"or"`
+	ResourcePaths []string           `json:"resource_paths"`
+	PayModels     []string           `json:"pay_models"`
+}
+
+type AuthRequestResponse struct {
+	Auth bool `json:"auth"`
+}
+
+/*
+	Authorization configuration parsing and validation
+*/
+
+func (authzConfig *AuthzConfig) UnmarshalJSON(data []byte) error {
+	/*
+		This custom unmarshal function allows us to parse the json config into the appropriate struct
+		depending on the `authz.version` value (eg struct AuthzVersion_0_1 when authz.version=0.1)
+	*/
+	var v interface{}
+	if err := json.Unmarshal(data, &v); err != nil {
+		return fmt.Errorf("could not parse 'authz' config: %v", err)
+	}
+	configObject := v.(map[string]interface{})
+
+	if configObject["version"] == nil {
+		// When there is no `authz` field, `authz.version` is unmarshalled to the default int
+		// value "0" which is interpreted as "no authz config".
+		// Here there is an `authz` field but no `authz.version`: assume it's a misconfiguration.
+		return fmt.Errorf("missing 'version' field in 'authz' config: %v", string(data))
+	} else {
+		authzConfig.Version = float32(configObject["version"].(float64))
+	}
+
+	// parse json data into the appropriate struct
+	if authzConfig.Version == 0.1 {
```

Comment:
nice, I like this a lot better, it's much more clear how to extend this in the future for a new version. to make it even better, a comment in the top-level struct with something like "add new versions here and adjust unmarshalling to the proper new type below" would be icing on the cake